pkgs/top-level/ocaml-packages.nix

```diff
@@ -1242,6 +1242,8 @@ let
       inherit (pkgs) unzip;
     };
 
+    opam-publish = callPackage ../development/ocaml-modules/opam-publish { };
```

Comment:
Does this really belongs to the `ocamlPackages` sets? i.e., does the behavior of the `opam-publish` binary depend on the version of OCaml that is used to build it?

Reply:
I don't think the behaviour changes in any way. So you'd recommend to move this package to top-level altogether, or to put an alias there?

I was wondering if it wasn't good to keep it in `ocamlPackages` so that if someone uses both packages from eg. `ocaml-ng.ocamlPackages_*` and `opam-publish` they can avoid duplicating the OCaml toolchain? I am just still discovering all of this, so sorry if I make little sense :p

Reply:
Then please move this out of `ocamlPackages`.

What do you call “duplicating the OCaml toolchain”? `opam-publish` does not use OCaml.

Reply:
> Then please move this out of `ocamlPackages`.

Happily. How does that work with the arguments of the package? Do I just take `ocamlPackages` as argument instead of the individual packages?

> What do you call “duplicating the OCaml toolchain”? `opam-publish` does not use OCaml.

In order to build the derivation, one would need to get the default OCaml toolchain as well as a bunch of packages from this toolchain. If the user in question is also using an alternative toolchain from `ocaml-ng.ocamlPackages_*`, then they are duplicating all of this work. (Of course, if things are in cache it makes very little difference.)

At runtime, I suppose `opam-publish` must need the OCaml runtime and some shared libraries, and all of that could be duplicated too, but that is probably not very big.

Reply:
As a reference this is how I did it for `ligo`: https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/compilers/ligo/default.nix

In practice `opam-publish` should be in the nix cache (built by hydra) so duplication shouldn't be much of a problem.

Reply:
> In order to build the derivation, one would need to get the default OCaml toolchain as well as a bunch of packages from this toolchain. If the user in question is also using an alternative toolchain from ocaml-ng.ocamlPackages_*, then they are duplicating all of this work.

If said user is really reluctant to download a pre-built binary of `opam-publish` from any cache, then they can simply override its derivation (as in `opam-publish.override { ocamlPackages = custom-trusted-ocaml-packages-set; }`).

Reply:
Fine. One more question: if I move the package to top-level, should I also move the corresponding Nix file? (from `development/ocaml-modules` to `development/tools` maybe? Thank you for your patience :-)

Reply:
Very nice. Thanks. May you squash it with the previous commit?

Reply:
Done. cf d94fa327d66.